### PR TITLE
Buoyancy force

### PR DESCRIPTION
**Pull request by shahabgol**
This PR adds the Buoyant force (using the Boussinesq approximation) to the Navier-Stokes assembler.
An application test (mms-buoyantforce_gls) is added.

### CODE REVIEW

File: include/solvers/navier_stokes_assemblers.h [hunk 1]
```diff
@@ -338,4 +338,47 @@ class LaplaceAssembly : public NavierStokesAssemblerBase<dim>
...
+   * @param scratch_data (see base class)
+   * @param copy_data (see base class)
+   */
+  virtual void
+  assemble_matrix(NavierStokesScratchData<dim> &        scratch_data,
+                  StabilizedMethodsTensorCopyData<dim> &copy_data) override;
+
+
+  /**
+   * @brief assemble_rhs Assembles the rhs
```
Comment (blaisb) [inline include/solvers/navier_stokes_assemblers.h:370]: Can you give a bit more detail? What is the term being assembled exactly?

File: include/solvers/navier_stokes_scratch_data.h [hunk 2]
```diff
@@ -517,6 +522,48 @@ class NavierStokesScratchData
       current_solution, fluid_pressure_gradients_at_particle_location);
   }
 
+
+  /**
+   * @brief enable_heat_transfer Enables the collection of the heat transfer data by the scratch
```
Comment (blaisb) [inline include/solvers/navier_stokes_scratch_data.h:527]:
```suggestion
   * @brief enable_heat_transfer Enables the collection of the heat transfer data (Temperature field) by the scratch
```

File: source/core/parameters.cc [hunk 3]
```diff
@@ -1159,7 +1170,7 @@ namespace Parameters
       prm.declare_entry(
         "type",
         "none",
-        Patterns::Selection("none|srf"),
+        Patterns::Selection("none|srf|gravity"),
```
Comment (blaisb) [inline source/core/parameters.cc:1173]:
```suggestion
        Patterns::Selection("none|srf"),
```

File: source/core/parameters_multiphysics.cc [hunk 4]
```diff
@@ -30,6 +30,11 @@ Parameters::Multiphysics::declare_parameters(ParameterHandler &prm)
                       Patterns::Bool(),
                       "Free surface calculation <true|false>");
 
+    prm.declare_entry("buoyant force",
```
Comment (blaisb) [inline source/core/parameters_multiphysics.cc:33]:
```suggestion
    prm.declare_entry("buoyancy force",
```
I think buoyancy force is the correct naming.

File: source/solvers/gls_navier_stokes.cc [hunk 5]
```diff
@@ -375,6 +376,13 @@ GLSNavierStokesSolver<dim>::setup_assemblers()
               this->simulation_parameters.velocity_sources));
         }
 
+      // Buoyant force
+      if (this->simulation_parameters.multiphysics.buoyant_force)
```
Comment (blaisb) [inline source/solvers/gls_navier_stokes.cc:380]:
```suggestion
      if (this->simulation_parameters.multiphysics.buoyancy_force)
```

File: source/solvers/gls_navier_stokes.cc [hunk 6]
```diff
@@ -519,6 +527,15 @@ GLSNavierStokesSolver<dim>::assemble_system_rhs()
                                        *this->mapping);
     }
 
+  if (this->simulation_parameters.multiphysics.buoyant_force)
```
Comment (blaisb) [inline source/solvers/gls_navier_stokes.cc:530]:
```suggestion
  if (this->simulation_parameters.multiphysics.buoyancy_force)
```

File: source/solvers/navier_stokes_assemblers.cc [hunk 7]
```diff
@@ -950,3 +950,59 @@ LaplaceAssembly<dim>::assemble_rhs(
...
+  // Loop over the quadrature points
+  for (unsigned int q = 0; q < n_q_points; ++q)
+    {
+      // Forcing term (gravity)
+      const Tensor<1, dim> force = scratch_data.force[q];
+
+      // Store JxW in local variable for faster access;
+      const double JxW = JxW_vec[q];
+
+      // Current and previous temperature values
```
Comment (blaisb) [inline source/solvers/navier_stokes_assemblers.cc:989]:
```suggestion
      // Current temperature values
```

File: source/solvers/navier_stokes_assemblers.cc [hunk 8]
```diff
@@ -950,3 +950,59 @@ LaplaceAssembly<dim>::assemble_rhs(
...
+      // NOT FROM SCRACH DATA BUT FROM HT SCRATCH DATA
```
Comment (blaisb) [inline source/solvers/navier_stokes_assemblers.cc:990]:
```suggestion
```

File: include/core/parameters_multiphysics.h [hunk 9]
```diff
@@ -38,6 +38,7 @@ namespace Parameters
     bool heat_transfer;
     bool tracer;
     bool free_surface;
+    bool buoyant_force;
```
Comment (blaisb) [inline include/core/parameters_multiphysics.h:41]:
```suggestion
    bool buoyancy_force;
```

**Review by blaisb — COMMENTED** [review]
Some very minor comments to address and this will be ready for merge.
I would try to find a better way to assess the validity of the test. LIke printing max and min velocity. Your analytical solution is only correct for the center of the geometry so it's not the best idea to monitor L2 error IMO

File: applications_tests/gls_navier_stokes_2d/mms-buoyantforce_gls.output [hunk 10]
```diff
@@ -0,0 +1,51 @@
+Running on 1 MPI rank(s)...
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply applications_tests/gls_navier_stokes_2d/mms-buoyantforce_gls.output:1]: corrected.

File: applications_tests/gls_navier_stokes_2d/mms-buoyantforce_gls.output [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+*****************************************************************
+   Number of active cells:       1024
+   Number of degrees of freedom: 3267
+   Volume of triangulation:      0.05
+   Number of thermal degrees of freedom: 1089
+L2 error velocity : 1.35526e-05
+L2 error temperature : 1.00606e-12
+
+*****************************************************************
+Steady iteration :        4/4
```
[the remark this answers is not in the document]
Comment (shahabgol) [reply applications_tests/gls_navier_stokes_2d/mms-buoyantforce_gls.output:34]: Changed to 2.

File: include/solvers/navier_stokes_assemblers.h (hunk 1 above)
Comment (shahabgol) [reply include/solvers/navier_stokes_assemblers.h:370]: Added.

**Comment by shahabgol** [comment]
I have applied the comments.
For assessing the validity of the test, I think comparison to analytical solution is good. It is true that the analytical solution is not correct at the top and bottom of the geometry, but the errors are still small for both velocity and temperature. What do you think?

**Comment by blaisb** [comment]
> I have applied the comments. For assessing the validity of the test, I think comparison to analytical solution is good. It is true that the analytical solution is not correct at the top and bottom of the geometry, but the errors are still small for both velocity and temperature. What do you think?

Okidoke, seems legit

**Review by blaisb — APPROVED** [review]
Ready for merge. Can you modify wiki to include parameters?

**Pull request merged**